src/atomate2/abinit/flows/dfpt.py

```diff
@@ -0,0 +1,224 @@
+    use_dde_sym: bool | None = False
+    dte_skip_permutations: bool | None = False
+    dte_phonon_pert: bool | None = False
+    dte_ixc: int | None = None
+
+    def make(
+        self,
+        structure: Structure | None = None,
+        restart_from: str | Path | None = None,
+    ) -> Flow:
+        """
+        Create a DFPT flow.
+
+        Parameters
+        ----------
+        structure : Structure
+            A pymatgen structure object.
+        restart_from : str or Path or None
+            One previous directory to restart from.
+
+        Returns
+        -------
+        Flow
+            A DFPT flow
+        """
+        static_job = self.static_maker.make(structure, restart_from=restart_from)
+        # To avoid metallic case=occopt=3 which is not okay wrt. DFPT \
+        # and occopt 1 with spin polarization requires spinmagntarget
+        static_job = update_factory_kwargs(
+            static_job, {"smearing": "nosmearing", "spin_mode": "unpolarized"}
+        )
+        # static_job = update_user_kpoints_settings(static_job,{"grid_density": 3000})
+        # static_job = update_user_abinit_settings( static_job,{'nstep': 500,
+        #                                                          'toldfe': 1e-22,
+        #                                                          'autoparal': 1,
+        #                                                          'npfft': 1,
+        #                                                          'chksymbreak': '0'})
+        static_job = update_user_kpoints_settings(static_job, {"grid_density": 1000})
+        static_job = update_user_abinit_settings(
+            static_job,
+            {
+                "nstep": 500,
+                "toldfe": 1e-6,
+                "autoparal": 1,
+                "npfft": 1,
+                "chksymbreak": "0",
+            },
+        )
+        jobs = [static_job]
+
+        if self.ddk_maker:
+            # generate the perturbations for the DDK calculations
+            ddk_perts = generate_ddk_perts(
+                gsinput=static_job.output.input.abinit_input,
+                use_symmetries=self.use_ddk_sym,
+            )
+            jobs.append(ddk_perts)
+
+            # perform the DDK calculations
+            ddk_calcs = run_ddk_rf(
+                perturbations=ddk_perts.output,
+                prev_outputs=static_job.output.dir_name,
+            )
+            jobs.append(ddk_calcs)
+
+        if self.dde_maker:
+            # generate the perturbations for the DDE calculations
+            dde_perts = generate_dde_perts(
+                gsinput=static_job.output.input.abinit_input,
+                use_symmetries=self.use_dde_sym,
+            )
+            jobs.append(dde_perts)
+
+            # perform the DDE calculations
+            dde_calcs = run_dde_rf(
+                perturbations=dde_perts.output,
+                prev_outputs=[[static_job.output.dir_name], ddk_calcs.output["dirs"]],
```

Comment:
Here the is no check, but of course if ddk are not calculated so should not be possible to get here. Maybe the best would be to check with a `__post_init__` that if the dde_maker is not None so should be the ddk_maker.

Reply:
Not sure I follow you. Do you mean that, if the ddk are not calculated, then it should stop before doing ```generate_ddk_perts```? Because in any case if the ddk fail then ```run_dde_rf``` is not started.

Reply:
In the current implementation if you do something like:

```python
flow = DfptFlowMaker(ddk_maker=None).make()
```

this be accepted, but will fail at runtime because when it gets to this like `ddk_calcs` will not be defined. Also you will get a puzzling erorr the the `ddk_calcs` variable is not defined. It is a bit pointless to wait the execution to fail as you can make this check when the `DfptFlowMaker`, for example something like this:

```python
def __post_init__(self):
    if self.dde_maker and not self.ddk_maker:
        raise ValueError("DDK caculations are required to DDE. Either provide a DDK Maker or remove the DDE one.")
```

The same checks can be done for DTE part with respect to DDK and DDE.

Now is relatively easy with "just" ddk, dde, dte, but when you add all the perturbations it may be helpful to avoid errors at runtime.

Reply:
Ah I see now. Indeed, it is useful to have this check.